cmd/maestro/server/grpc_broker.go

```diff
@@ -0,0 +1,532 @@
+		creds, err := credentials.NewServerTLSFromFile(config.TLSCertFile, config.TLSKeyFile)
+		if err != nil {
+			glog.Fatalf("Failed to generate credentials %v", err)
+		}
+		grpcServerOptions = append(grpcServerOptions, grpc.Creds(creds))
+		glog.Infof("Serving gRPC broker with TLS at %s", config.BrokerBindPort)
+	} else {
+		glog.Infof("Serving gRPC broker without TLS at %s", config.BrokerBindPort)
+	}
+
+	return &GRPCBroker{
+		grpcServer:         grpc.NewServer(grpcServerOptions...),
+		resourceService:    env().Services.Resources(),
+		statusEventService: env().Services.StatusEvents(),
+		bindAddress:        env().Config.HTTPServer.Hostname + ":" + config.BrokerBindPort,
+		subscribers:        make(map[string]*subscriber),
+	}
+}
+
+// Start starts the gRPC broker
+func (bkr *GRPCBroker) Start(ctx context.Context) {
+	lis, err := net.Listen("tcp", bkr.bindAddress)
+	if err != nil {
+		glog.Fatalf("failed to listen: %v", err)
+	}
+	pbv1.RegisterCloudEventServiceServer(bkr.grpcServer, bkr)
+	go func() {
+		if err := bkr.grpcServer.Serve(lis); err != nil {
+			glog.Fatalf("failed to start gRPC broker: %v", err)
+		}
+	}()
+
+	// wait until context is canceled
+	<-ctx.Done()
+	log.Infof("Shutting down gRPC broker")
+}
+
+// Publish in stub implementation for maestro agent publish resource status back to maestro server.
+func (bkr *GRPCBroker) Publish(ctx context.Context, pubReq *pbv1.PublishRequest) (*emptypb.Empty, error) {
+	// WARNING: don't use "evt, err := pb.FromProto(pubReq.Event)" to convert protobuf to cloudevent
+	evt, err := binding.ToEvent(ctx, grpcprotocol.NewMessage(pubReq.Event))
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert protobuf to cloudevent: %v", err)
+	}
+
+	eventType, err := types.ParseCloudEventsType(evt.Type())
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cloud event type %s, %v", evt.Type(), err)
+	}
+
+	glog.V(4).Infof("receive the event with grpc, %s", evt)
+
+	// handler resync request
+	if eventType.Action == types.ResyncRequestAction {
+		err := bkr.respondResyncSpecRequest(ctx, eventType.CloudEventsDataType, evt)
+		if err != nil {
+			return nil, fmt.Errorf("failed to respond resync spec request: %v", err)
+		}
+		return &emptypb.Empty{}, nil
+	}
+
+	resourceID, err := cetypes.ToString(evt.Extensions()[types.ExtensionResourceID])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resourceid extension: %v", err)
+	}
+
+	consumerName, err := cetypes.ToString(evt.Extensions()[types.ExtensionClusterName])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get clustername extension: %v", err)
+	}
+
+	found, svcErr := bkr.resourceService.Get(ctx, resourceID)
+	if svcErr != nil {
+		return nil, fmt.Errorf("failed to get resource %s, %s", resourceID, svcErr.Error())
+	}
+
+	if found.ConsumerName != consumerName {
+		return nil, fmt.Errorf("unmatched consumer name %s for resource %s", consumerName, resourceID)
+	}
+
+	specEvent, err := api.JSONMAPToCloudEvent(found.Payload)
```

Comment:
is the following part same with RESTful part? can we make a common function to handle this?

Reply:
Yes, I created a common function (handleStatusUpdate) so we can handle the status event from either the MQTT or gRPC broker using the same logic.